desktop_version/src/DiscordNetwork.c

```diff
@@ -0,0 +1,162 @@
+
+
+
+#define FOREACH_FUNC(rettype, name, params) static rettype (*name) params = NULL;
+FUNC_LIST
+#undef FOREACH_FUNC
+
+static void ClearPointers(void)
+{
+    SDL_UnloadObject(libHandle);
+    libHandle = NULL;
+#define FOREACH_FUNC(rettype, name, params) name = NULL;
+    FUNC_LIST
+#undef FOREACH_FUNC
+}
+
+void DISCORD_shutdown(void)
+{
+    if (libHandle)
+    {
+        ClearPointers();
+    }
+}
+
+int32_t DISCORD_init(void)
+{
+#if defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__HAIKU__) || defined(__DragonFly__)
+    return 0;
+#endif
+    libHandle = SDL_LoadObject(DISCORD_LIBRARY);
+    if (!libHandle)
+    {
+        vlog_info(DISCORD_LIBRARY " not found!");
+        printf("Can't load object %s : %s\n", DISCORD_LIBRARY, SDL_GetError());
+        return 0;
+    }
+
+#define FOREACH_FUNC(rettype, name, params) \
+    name = (rettype (*) params) (intptr_t) SDL_LoadFunction(libHandle, #name); \
+    if (!name) \
+    { \
+        vlog_error(DISCORD_LIBRARY " symbol " #name " not found!"); \
+        ClearPointers(); \
+        return 0; \
+    }
+    FUNC_LIST
+#undef FOREACH_FUNC
+    memset(&app, 0, sizeof(app));
+
+    struct DiscordCreateParams params;
+    params.client_id = DISCORD_CLIENT_ID;
+    params.flags = DiscordCreateFlags_NoRequireDiscord;
+
+    if(DiscordCreate(DISCORD_VERSION, &params, &app.core) != DiscordResult_Ok)
+    {
+        discordDetected = false;
+        return 0;
+    }
+
+    if(app.core == NULL)
+    {
+        discordDetected = false;
+        return 0;
+    }
+
+
+    SDL_strlcpy(activity.assets.large_image, "vvvvvv", sizeof(activity.assets.large_image));
+    SDL_strlcpy(activity.assets.large_text, "Outside Dimension VVVVVV", sizeof(activity.assets.large_text));
+
+    app.activityMan = app.core->get_activity_manager(app.core);
+    app.activityMan->update_activity(app.activityMan, &activity, NULL, NULL);
+
+
+    discordDetected = true;
+    return 1;
+}
+
+bool DISCORD_REQUIRE(int x) {
+    if(!discordDetected)
+    {
+        return false;
+    }
+    if(x != DiscordResult_Ok)
+    {
+        return false;
+    }
+    return true;
+}
+
+void DISCORD_update(const char* area, const char* roomname)
+{
+    if(app.core == NULL) {
+        return;
+    }
+    if(app.activityMan == NULL)
+    {
+        app.activityMan = app.core->get_activity_manager(app.core);
+    }
+    if(activity.state == roomname || activity.assets.large_text == area)
+    {
+        DISCORD_REQUIRE(app.core->run_callbacks(app.core));
+        return;
+    }
+    SDL_strlcpy(activity.state, roomname, sizeof(activity.state));
+    SDL_strlcpy(activity.assets.large_image, "vvvvvv", sizeof(activity.assets.large_image));
+    SDL_strlcpy(activity.assets.large_text, area, sizeof(activity.assets.large_text));
+
+    app.activityMan->update_activity(app.activityMan, &activity, NULL, NULL);
+
+    DISCORD_REQUIRE(app.core->run_callbacks(app.core));
```

Comment:
I think how the update function should work is something like this:

```cpp
if (app.core->run_callbacks(app.core) != DiscordResult_Ok)
{
    // No Discord
    return;
}

// It was Ok, so we can now do things with Discord
if (activity is outdated)
{
    update activity;
}
```

Also, with C strings, if you want to check that two strings are equal, you need to do `SDL_strcmp(a, b) == 0` instead of `a == b`. `activity.state == roomname` and `activity.assets.large_text == area` will check that the two strings are stored in exactly the same bit of memory, rather than the strings having the same text.

Reply:
@Daaaav I did consider this, but you are meant to run the run_callbacks function after the activity update (source as per Discord C example)

Reply:
I don't think it's necessarily that you need to run `run_callbacks` after the activity update; the example code just wanted to demonstrate updating the activity only one time and then run `run_callbacks` in a main loop, as a simplified example. And even if `run_callbacks` is the thing that "confirms" an activity update... well, then it's only going to be delayed by one frame.

The main advantage of switching them around is that we don't have to bother constantly updating the activity and telling Discord about it, if we know Discord isn't ready to take it anyway.

Reply:
Alright, in my opinion 1 frame delay is way too much (lol average gamer) but I can switch it around if you really want, I'll leave it as a single commit incase anyone wants to rollback.

Reply:
Wait... I don't actually thing we should do that. We need to handle for app.core == NULL not running a callback, as if app.core was to be NULL then we would have a segfault. So instead of replacing error handling with that I'll just add it on.

Reply:
The commit fucked something up - it no longer works. Reverting.

Reply:
I put some debug logs in for different scenarios:

[ERROR] Discord API failed to initialise!

Reply:
[ERROR] DiscordResult_Ok
[ERROR] Discord API failed to initialise!
even weirder... probably an easy fix

Reply:
Take your time to get it fully working - you don't have to post a message every step along the way 😛

Reply:
Fixed it, just incorrect handling of DISCORD_REQUIRE, I fixed that but introduced another variable. I have to do it unless of course you want discordDetected to be true by default, which would be weird.